activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -53,8 +53,10 @@ def initialize(connection, logger, connection_options, config)
 
         @statements = StatementPool.new(self.class.type_cast_config_to_integer(config[:statement_limit]))
 
-        if version < "5.1.10"
-          raise "Your version of MySQL (#{version_string}) is too old. Active Record supports MySQL >= 5.1.10."
+        if version < "5.7.9"
+          raise "Your version of MySQL (#{version_string}) is too old. Active Record supports MySQL >= 5.7.9."
+        elsif mariadb? && version < "10.2.2"
+          raise "Your version of MariaDB (#{version_string}) is too old. Active Record supports MariaDB >= 10.2.2"
```

Comment:
Thanks for the comment. 

I still wanted to bump the minimum version of MySQL and MariaDB to support utf8mb4 character set without any my.cnf changes.

I believe many users have successfully migrated to utf8mb4 character set using MySQL 5.5 and 5.6.
Also, there is plenty number of users who just configured utf8mb4 character set and have been struggling with "ERROR 1071 (42000): Specified key was too long; max key length is 767 bytes". i.e. Googled "utf8mb4 Specified key was too long" and found more than 10,000 articles.

To address this error with MySQL 5.5 and 5.6, my.cnf needs update to change `innodb_file_per_table`, `innodb_file_format` and `innodb_large_prefix` to non-default value. MySQL 5.7.9 and MariaDB 10.2.2 introduces `innodb_default_row_format` to allow longer 3072 byte key size by default. No need to change my.cnf at all to support utf8mb4 character set.

I think the biggest advantage to bump MySQL version to 5.7 and MariaDB 10.2.2 is no need to change my.cnf at all to support utf8mb4 character set. 

I might have misunderstood your point. Further discussions are appreciated.

Reply:
Thanks for the feedback. I think I get points of review from Jeremy and Rafael and still, I'd like to keep this pull request as it is for now.

Because without this pull request users can use utf8mb4 with MySQL 5.5 with some necessary changes in my.cnf. I am thinking that the main point of this pull request is "by default". 

Upgrading MySQL server version is not an easy task, 
but still, it is a good timing bump the minimum requirement of MySQL when Rails major version bumps from 5 to 6.

I'd like to see some more feedback if many of these feedback prefers not to bump the minimum version of MySQL I'll revise or close this pull request.

Thanks again for the useful feedback.

Reply:
@matthewd @tenderlove can you give us your opinion about this?

Reply:
@yahonda Agreed that Rails 6 apps should have correct "by default" utf-8 charset behavior with MySQL/MariaDB and that bumping the required db version is a strong way to guarantee it.

I suggest we allow existing deployed apps to upgrade to Rails 6 if they have taken responsibility for compatibility. For example, if they have already migrated to utf8mb4 on an AWS Aurora cluster with MySQL 5.6, then they already set `encoding: utf8mb4` in config/database.yml and we can trust their configuration.

So perhaps we can require modern databases by default, but also allow "migrated" databases:
* New Rails 6 apps require modern MySQL/MariaDB.
* Upgrading Rails 6 apps that do not specify encoding in database.yml, and hence inherit the new default encoding, also require modern MySQL/MariaDB.
* Upgrading Rails 6 apps that explicitly specify encoding in database.yml are assumed to be using a compatible, properly configured database.

I think we could do this with a version check in `initialize` that's only triggered when `!config.include?(:encoding)`.

Reply:
To be clear, I'm not suggesting closing this PR either! This change is very desirable, and so is an "escape hatch" for existing app deployments that are already compatible.

If there are other MySQL/MariaDB features we'd like to rely on, or old database support we'd like to drop from the adapter, we can certainly update our minimum version requirements for those reasons as well.

Reply:
I think I get lost on the version interactions here... if new/recent MySQL/MariaDB has good default behaviour for encoding, and older versions _can_ be configured to have the good behaviour (but might not be), and if you're using an old version that _isn't_ configured well then everything will still work fine but you'll get an error from some DDL if you aren't overriding default lengths (is that right?)... then I'm not sure we need to do anything.

Bumping our compatible server versions just to enforce "good encodings are good" seems pretty aggressive to me. It may not be our recommended configuration, but I'd want compelling internal driver benefits to warrant outright refusing to talk to peoples databases.

You _should_ use `utf8mb4`, which is also what you get automatically from a modern DB server... but I don't see why even a brand new application shouldn't still work with an arbitrarily-encoded database, if the only implication is that you might have encoding issues. There are plenty of app-external things you _should_ be doing, but it's not our job to enforce them. ("The kernel version you're using has a known bug in the ext4 driver; refusing to boot Rails"?)

Reply:
Thanks for additional feedbacks. I'll update my pull request to relax database version requirements.
I need to take a look at the differences between `charset` and `encoding` first.